### PR TITLE
feat: Use NSSM for containerd and collect containerd logs

### PR DESCRIPTION
**Pull request by jsturtevant**
**Reason for Change**:

We need to collect the containerd logs for analysis.

**Issue Fixed**:

**Credit Where Due**:
This WPR file was shared via @kevpar on slack

Does this change contain code from or inspired by another project?

- [x] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [x] No
- [ ] Yes

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jsturtevant** [comment]
@kevpar any suggestions on the wprp profile?

File: parts/k8s/windowscontainerdfunc.ps1 [hunk 1]
```diff
@@ -94,7 +94,9 @@ function Enable-Logging {
     & $diag -Start -ProfilePath "$global:ContainerdInstallLocation\ContainerPlatform.wprp!ContainerPlatformPersistent" -TempPath $logs
   }
   else {
-    Write-Log "Containerd hyperv logging script not avalaible"
+    Write-Log "Containerd logging enabled using $global:ContainerdInstallLocation\containerd.wprp"
+    DownloadFileOverHttp -Url 'https://raw.githubusercontent.com/jsturtevant/aks-engine/log-containerd-containers-running/scripts/containerd.wprp' -DestinationPath "$global:ContainerdInstallLocation\containerd.wprp"
```
Comment (jsturtevant) [inline parts/k8s/windowscontainerdfunc.ps1:98]: using my branch for testing

**Comment by jsturtevant** [comment]
This is working (truncated output). Tests after a node restart as well.

```
Compressing all logs to 2018k8s01000000-20210204-115313_logs.zip 

C:\Users\azureuser\AppData\Local\Temp\20210204-115313-containerd.txt
C:\Users\azureuser\AppData\Local\Temp\20210204-115313-containerd.etl




    Directory: C:\Users\azureuser


Mode          LastWriteTime Length Name
----          ------------- ------ ----
-a----   2/4/2021  11:53 PM 199260 2018k8s01000000-20210204-115313_logs.zip
```

note: if there is an unplanned node crashed, logs prior to the crash will not be retained. If required a logging agent should be configured to listen to the ETW events and persist longer term.

**Comment by jsturtevant** [comment]
and the etl file:

image

**Comment by jsturtevant** [comment]
Discussed this further and we will not be turning the scripts on by default.  We will provide a script to turn on the logging when needed and if folks want to collect logging for containerd full time they will need to provide a logging agent that listens to the etw events.

holding for updates to not turn this on by defualt
/hold

File: scripts/collect-windows-logs.ps1 [hunk 2]
```diff
@@ -86,6 +86,18 @@ else {
   Write-Host "Containerd hyperv logs not avalaible"
 }
 
+# log containerd containers (this is done for docker via networking collectlogs.ps1)
+$res = Get-Command ctr.exe -ErrorAction SilentlyContinue
+if ($res)
+{ 
+  & ctr.exe -n k8s.io c ls > "$ENV:TEMP\$timeStamp-containerd-containers.txt"
+  & ctr.exe -n k8s.io t ls > "$ENV:TEMP\$timeStamp-containerd-tasks.txt"
+  & wpr.exe -stop "$ENV:TEMP\$timeStamp-containerd.etl"
```
Comment (mainred) [inline scripts/collect-windows-logs.ps1:95]: New to wpr, just curious, where do we start the trace?

Comment (jsturtevant) [reply scripts/collect-windows-logs.ps1:95]: We aren't starting the trace.  Someone would need to start the trace, repo the issue and then run this script to collect all the logs.  If there is no trace running this would report that, and continue on.  We are going to try a different approach all together here so this will change.

File: parts/k8s/containerdtemplate.toml [hunk 3]
```diff
@@ -32,6 +32,7 @@ state = "C:\\ProgramData\\containerd\\state"
     max_container_log_line_size = 16384
     [plugins.cri.containerd]
       snapshotter = "windows"
+      discard_unpacked_layers = true
```
Comment (marosset) [inline parts/k8s/containerdtemplate.toml:35]: Is this setting related to logging?

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 4]
```diff
@@ -256,6 +256,18 @@ try
 
         Write-KubeClusterConfig -MasterIP $MasterIP -KubeDnsServiceIp $KubeDnsServiceIp
 
+        Write-Log "Download kubelet binaries and unzip"
+        Get-KubePackage -KubeBinariesSASURL $global:KubeBinariesPackageSASURL
+
+        # this overwrite the binaries that are download from the custom packge with binaries
```
Comment (marosset) [inline parts/k8s/kuberneteswindowssetup.ps1:262]: I think the first line is not needed in the comment section (i know you are just moving it around but wouldn't hurt to clean it up)

File: pkg/engine/templates_generated.go [hunk 5]
```diff
@@ -18854,24 +18855,42 @@ func k8sWindowsconfigfuncPs1() (*asset, error) {
...
+  # setup containerd
+  & "$KubeDir\nssm.exe" install containerd $global:Containerdbinary | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd AppDirectory $KubeDir | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd DisplayName containerd | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd Description containerd | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd Start SERVICE_DEMAND_START | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd ObjectName LocalSystem | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd Type SERVICE_WIN32_OWN_PROCESS | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd AppThrottle 1500 | RemoveNulls
+  & "$KubeDir\nssm.exe" set containerd AppStdout C:\k\containerd.log | RemoveNulls
```
Comment (marosset) [inline pkg/engine/templates_generated.go:18884]: Should we log to the same directory as containerd or to c:\k\?
If we log to c:\k\ logs will all be together but if we log next to containerd (c:\program files\contaienrd) it might be less confusing for someone looking for the logs...

Comment (marosset) [reply pkg/engine/templates_generated.go:18884]: Also, should we use $KubeDir instead of c:\k\?

**Review by marosset — COMMENTED** [review]
mostly looks good, had a few minor comments tho.

File: parts/k8s/containerdtemplate.toml (hunk 3 above)
Comment (jsturtevant) [reply parts/k8s/containerdtemplate.toml:35]: no. was given guidance to use this setting in k8s env which will help with disk space. I could submit it as separate pr if we want.

File: pkg/engine/templates_generated.go (hunk 5 above)
Comment (jsturtevant) [reply pkg/engine/templates_generated.go:18884]: I was leaning towards `c:\k` but don't have strong feelings.  I think the hardcoded `c:\k` was copy past from testing. I will update that.

Comment (jsturtevant) [reply pkg/engine/templates_generated.go:18884]: I set the working dir to c:\k as well to fix the azure cni logging issue

File: parts/k8s/containerdtemplate.toml (hunk 3 above)
Comment (marosset) [reply parts/k8s/containerdtemplate.toml:35]: nah, just curious

File: vhd/packer/configure-windows-vhd.ps1 [hunk 6]
```diff
@@ -159,19 +160,6 @@ function Install-ContainerD {
     $newPath = [Environment]::GetEnvironmentVariable("Path", [EnvironmentVariableTarget]::Machine) + ";$installDir"
     [Environment]::SetEnvironmentVariable("Path", $newPath, [EnvironmentVariableTarget]::Machine)
     $env:Path += ";$installDir"
-
-    Write-Log "Registering containerd as a service"
-    & containerd.exe --register-service
```
Comment (marosset) [inline vhd/packer/configure-windows-vhd.ps1:164]: We still need to install containerd in the VHD in order to pull nanoserver/servercore/pause image

Comment (marosset) [reply vhd/packer/configure-windows-vhd.ps1:164]: or at least start it

Comment (jsturtevant) [reply vhd/packer/configure-windows-vhd.ps1:164]: make sense. I was hoping we wouldn't need to duplicate the configuration set up but I guess we will need that

Comment (marosset) [reply vhd/packer/configure-windows-vhd.ps1:164]: You can probably get by with a `Start-Job` and not need to register this as a service.
This won't persist across reboots tho

Comment (marosset) [reply vhd/packer/configure-windows-vhd.ps1:164]: Also, i wonder why the Windows VHD CI job didn't trigger here...

Comment (jsturtevant) [reply vhd/packer/configure-windows-vhd.ps1:164]: I've made this update. PTAL

File: scripts/collect-windows-logs.ps1 (hunk 2 above)
Comment (jsturtevant) [reply scripts/collect-windows-logs.ps1]: We moved to using nssm to enable consistent logging story across our components

File: vhd/packer/configure-windows-vhd.ps1 (hunk 6 above)
Comment (marosset) [reply vhd/packer/configure-windows-vhd.ps1:164]: Looks good.
Let's make sure the containerD VHD job passes the CI job!

**Review by marosset — APPROVED** [review]
